atc/exec/set_pipeline_step.go

```diff
@@ -0,0 +1,243 @@
+	if !diffExists {
+		logger.Debug("no-diff")
+
+		fmt.Fprintf(stdout, "No diff found.\n")
+		step.succeeded = true
+		step.delegate.Finished(logger, true)
+		return nil
+	}
+
+	fmt.Fprintf(stdout, "Updating the pipeline.\n")
+	pipeline, _, err = team.SavePipeline(step.plan.Name, *atcConfig, fromVersion, false)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(stdout, "Done successfully.\n")
+	logger.Info("saved-pipeline", lager.Data{"team": team.Name(), "pipeline": pipeline.Name()})
+	step.succeeded = true
+	step.delegate.Finished(logger, true)
+
+	return nil
+}
+
+func (step *SetPipelineStep) Succeeded() bool {
+	return step.succeeded
+}
+
+type setPipelineSource struct {
+	ctx    context.Context
+	logger lager.Logger
+	repo   *artifact.Repository
+	step   *SetPipelineStep
+}
+
+// streamInBytes streams a file from other resource and returns a byte array.
+func (s setPipelineSource) streamInBytes(path string) ([]byte, error) {
+	stream, err := s.repo.StreamFile(s.ctx, s.logger, path)
+	if err != nil {
+		return nil, err
+	}
+	defer stream.Close()
+
+	byteConfig, err := ioutil.ReadAll(stream)
+	if err != nil {
+		return nil, err
+	}
+
+	return byteConfig, nil
+}
+
+// FetchConfig streams pipeline configure file and var files from other resources
+// and construct an atc.Config object.
+func (s setPipelineSource) FetchConfig() (*atc.Config, error) {
+	config, err := s.streamInBytes(s.step.plan.File)
+	if err != nil {
+		return nil, err
+	}
+
+	staticVarss := []vars.Variables{}
+	if len(s.step.plan.Vars) > 0 {
+		staticVarss = append(staticVarss, vars.StaticVariables(s.step.plan.Vars))
+	}
+	for _, lvf := range s.step.plan.VarFiles {
+		bytes, err := s.streamInBytes(lvf)
+		if err != nil {
+			return nil, err
+		}
+
+		sv := vars.StaticVariables{}
+		err = yaml.Unmarshal(bytes, &sv)
+		if err != nil {
+			return nil, err
+		}
+
+		staticVarss = append(staticVarss, sv)
+	}
+
+	if len(staticVarss) > 0 {
+		config, err = vars.NewTemplateResolver(config, staticVarss).Resolve(false, false)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	atcConfig := atc.Config{}
+	err = yaml.Unmarshal(config, &atcConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &atcConfig, nil
+}
+
+func (s setPipelineSource) Validate() error {
+	if s.step.plan.File == "" {
+		return errors.New("file is not specified")
+	}
+
+	return nil
+}
+
+func (s setPipelineSource) EvaluatePlan(variables vars.Variables) error {
+	params := atc.Params{
+		"file":      s.step.plan.File,
+		"vars":      s.step.plan.Vars,
+		"var_files": s.step.plan.VarFiles,
+	}
+
+	params, err := creds.NewParams(variables, params).Evaluate()
+	if err != nil {
+		return err
+	}
+
+	s.step.plan.File = params["file"].(string)
+	s.step.plan.Vars = params["vars"].(map[string]interface{})
+	s.step.plan.VarFiles = []string{}
+	for _, ele := range params["var_files"].([]interface{}) {
+		s.step.plan.VarFiles = append(s.step.plan.VarFiles, ele.(string))
+	}
+
+	if s.step.plan.Name == "self" {
```

Comment:
Hmm I guess I was hoping this could be done in `atc/scheduler/factory/build_factory.go` so the `atc.Plan` object is always verbatim (no "self" magic), but it looks like that object doesn't have access to the pipeline name. Maybe it could be passed in?

I think this is a little important because `atc.Plan` types are meant to be fairly "dry" and can be submitted outside of the context of a pipeline, like with `fly execute`. So conceptually it makes sense to me for `self` to be pipeline-config magic that gets de-sugared into a regular build plan.

Reply:
@vito I just changed the first argument of `BuildFactory.Create` from `JobConfig` to `Job`, so that we don't add extra parameter and we can figure out pipeline name from job.